### PR TITLE
Support subscription payments

### PR DESCRIPTION
**Pull request by BushraAsif**

### CODE REVIEW

File: Model/Gateway.php [hunk 1]
```diff
@@ -387,6 +388,10 @@ private function preparePaymentRequest($order, $orderLines, $orderId, $terminalI
                 $request->setLanguage($langArr[0]);
             }
         }
+        $quote = $this->quote->loadByIdWithoutStore($order->getQuoteId());
+        if($this->validateQuote($quote)){
```
Comment (AJenbo) [inline Model/Gateway.php:392]: Inconsistent code style
```suggestion
        if ($this->validateQuote($quote)) {
```

File: Observer/CaptureObserver.php [hunk 2]
```diff
@@ -262,24 +263,31 @@ private function shippingTrackingInfo($invoice)
...
-        $api->setOrderLines($orderLines);
-        $shippingTrackingInfo = $this->shippingTrackingInfo($invoice);
-        /*send shipping tracking info if exists*/
-        if (!empty($shippingTrackingInfo)) {
-            $api->setTrackingInfo($shippingTrackingInfo);
-        }
+        if ($paymentType === 'subscription') {
+            $api = new ChargeSubscription($this->systemConfig->getAuth($storeCode));
+            $api->setTransaction($payment->getLastTransId());
+            $api->setAmount((float)number_format($invoice->getGrandTotal(), 2, '.', ''));
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:275]: If you are casting the result of number_format (a `string`) to a `float` then you are doing something wrong.

```suggestion
            $api->setAmount(round($invoice->getGrandTotal(), 2));
```

File: Observer/CaptureObserver.php [hunk 3]
```diff
@@ -262,24 +263,31 @@ private function shippingTrackingInfo($invoice)
...
+        } else {
+            $api = new CaptureReservation($this->systemConfig->getAuth($storeCode));
+            if ($invoice->getTransactionId()) {
+                $api->setInvoiceNumber($invoice->getTransactionId());
+            }
+            $api->setAmount((float)number_format($invoice->getGrandTotal(), 2, '.', ''));
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:281]:
```suggestion
            $api->setAmount(round($invoice->getGrandTotal(), 2));
```

File: Observer/CaptureObserver.php [hunk 4]
```diff
@@ -262,24 +263,31 @@ private function shippingTrackingInfo($invoice)
...
+            $api->setOrderLines($orderLines);
+            $shippingTrackingInfo = $this->shippingTrackingInfo($invoice);
+            /*send shipping tracking info if exists*/
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:284]:
```suggestion
            // Send shipping tracking info if exists
```

File: Observer/CaptureObserver.php [hunk 5]
```diff
@@ -262,24 +263,31 @@ private function shippingTrackingInfo($invoice)
...
+            if (!empty($shippingTrackingInfo)) {
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:285]: What type is $shippingTrackingInfo?

File: Response/ChargeSubscriptionResponse.php [hunk 6]
```diff
@@ -25,4 +25,14 @@
 
 class ChargeSubscriptionResponse extends ReservationOfFixedAmountResponse
 {
+    public $CaptureResult;
```
Comment (AJenbo) [inline Response/ChargeSubscriptionResponse.php:28]:
```suggestion
    /** @var float */
    public $CaptureResult;
```

File: Model/Gateway.php [hunk 7]
```diff
@@ -463,4 +468,57 @@ public function fixedProductTax($order){
...
+        $buyRequest = $this->getBuyRequestObject($item);
+
+        return $buyRequest->getData('subscribe') === 'subscribe';
+    }
+    /**
+     * @param CartInterface $quote
+     *
+     * @return bool
+     */
+    public function validateQuote($quote): bool
```
Comment (AJenbo) [inline Model/Gateway.php:509]: The readme mentions PHP 5.5 support but this will require PHP 7.0. Have the requirements changed but not been updated?

Comment (AJenbo) [reply Model/Gateway.php:509]: I see that it's supposed to support 5.6 now, so this is defiantly an issue.

File: Response/ChargeSubscriptionResponse.php (hunk 6 above)
Comment (AJenbo) [reply Response/ChargeSubscriptionResponse.php:28]: Looks like this was copied from an unapproved version of API-PHP, probably best to have that reviewed and approved first.l

File: Observer/CaptureObserver.php (hunk 5 above)
Comment (BushraAsif) [reply Observer/CaptureObserver.php]: The type of this variable is an array. I have removed the check

File: Model/Gateway.php (hunk 7 above)
Comment (BushraAsif) [reply Model/Gateway.php]: Removed

File: Observer/CaptureObserver.php [hunk 8]
```diff
@@ -262,24 +263,29 @@ private function shippingTrackingInfo($invoice)
      * @param $orderObject
      * @param $payment
      * @param $storeCode
+     * @param $paymentType
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:266]: Missing type

File: Observer/CaptureObserver.php [hunk 9]
```diff
@@ -262,24 +263,29 @@ private function shippingTrackingInfo($invoice)
...
-        $api = new CaptureReservation($this->systemConfig->getAuth($storeCode));
-        if ($invoice->getTransactionId()) {
-            $api->setInvoiceNumber($invoice->getTransactionId());
-        }
-        $api->setAmount((float)number_format($invoice->getGrandTotal(), 2, '.', ''));
-        $api->setOrderLines($orderLines);
-        $shippingTrackingInfo = $this->shippingTrackingInfo($invoice);
-        /*send shipping tracking info if exists*/
-        if (!empty($shippingTrackingInfo)) {
+        $grandTotal = (float)$invoice->getGrandTotal();
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:272]: What type is `getGrandTotal` returning?

File: Observer/CaptureObserver.php (hunk 5 above)
Comment (AJenbo) [reply Observer/CaptureObserver.php]: Ok as long as setTrackingInfo can take an empty array :)

File: Observer/CaptureObserver.php [hunk 10]
```diff
@@ -108,18 +108,19 @@ public function execute(\Magento\Framework\Event\Observer $observer)
...
             $orderLines = $this->processInvoiceOrderLines($invoice);
             //Send request for payment refund
-            $this->sendInvoiceRequest($invoice, $orderLines, $orderObject, $payment, $storeCode);
+            $this->sendInvoiceRequest($paymentType, $invoice, $orderLines, $orderObject, $payment, $storeCode);
         }
     }
 
     /**
-     * @param $invoice
+     * @param null|object $invoice
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:123]: If `null` was given here the application would crash on line 129, the correct object appears to be `Magento\Sales\Model\Order\Invoice`:
```suggestion
     * @param Magento\Sales\Model\Order\Invoice $invoice
```

File: Observer/CaptureObserver.php [hunk 11]
```diff
@@ -158,9 +159,9 @@ private function processInvoiceOrderLines($invoice)
     }
 
     /**
-     * @param $couponCodeAmount
-     * @param $invoice
-     * @param $discountAllItems
+     * @param int|float $couponCodeAmount
+     * @param null|object $invoice
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:163]:
```suggestion
     * @param float $couponCodeAmount
     * @param Magento\Sales\Model\Order\Invoice $invoice
```

File: Observer/CaptureObserver.php [hunk 12]
```diff
@@ -234,7 +235,7 @@ private function itemOrderLines($couponCodeAmount, $invoice, $discountAllItems)
     }
 
     /**
-     * @param $invoice
+     * @param null|object $invoice
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:238]:
```suggestion
     * @param Magento\Sales\Model\Order\Invoice $invoice
```

File: Observer/CaptureObserver.php [hunk 13]
```diff
@@ -257,29 +258,34 @@ private function shippingTrackingInfo($invoice)
...
-     * @param $invoice
-     * @param $orderLines
-     * @param $orderObject
-     * @param $payment
-     * @param $storeCode
+     * @param null|object $invoice
+     * @param array $orderLines
+     * @param null|object $orderObject
+     * @param array $payment
+     * @param int|string $storeCode
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:265]:
```suggestion
     * @param Magento\Sales\Model\Order\Invoice $invoice
     * @param array $orderLines
     * @param Magento\Sales\Model\Order $orderObject
     * @param array $payment
     * @param string $storeCode
```

File: Observer/CaptureObserver.php [hunk 14]
```diff
@@ -315,7 +321,7 @@ private function sendInvoiceRequest($invoice, $orderLines, $orderObject, $paymen
     }
 
     /**
-     * @param $order
+     * @param null|object $invoice
      *
      * @return float|int
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:326]:
```suggestion
     * @param Magento\Sales\Model\Order\Invoice $invoice
     *
     * @return float
```

Maybe also change `$weeTaxAmount = 0;` to `$weeTaxAmount = 0.0;`

File: Response/ChargeSubscriptionResponse.php [hunk 15]
```diff
@@ -25,4 +25,16 @@
 
 class ChargeSubscriptionResponse extends ReservationOfFixedAmountResponse
 {
+    /** @var float */
+    public $CaptureResult;
+    
+    /**
+     * @param float $CaptureAmount
+     * @return CaptureReservationResponse
```
Comment (AJenbo) [inline Response/ChargeSubscriptionResponse.php:33]:
```suggestion
     * @return $this
```

File: Observer/CaptureObserver.php (hunk 9 above)
Comment (BushraAsif) [reply Observer/CaptureObserver.php:272]: It's returning string

Comment (AJenbo) [reply Observer/CaptureObserver.php:272]: on, round should do an implicit cast, but doing an explicit one is also fine

File: Observer/CaptureObserver.php [hunk 16]
```diff
@@ -315,13 +321,13 @@ private function sendInvoiceRequest($invoice, $orderLines, $orderObject, $paymen
     }
 
     /**
-     * @param $order
+     * @param Magento\Sales\Model\Order\Invoice $invoice
      *
      * @return float|int
```
Comment (AJenbo) [inline Observer/CaptureObserver.php:326]:
```suggestion
     * @return float
```

**Review by AJenbo — APPROVED** [review]

**Pull request merged**